doc/hntool.1
.TH HNTOOL 1 local
.SH NAME
hntool \- Hardening tool for Unix.
.SH SYNOPSIS
.ll +8
.B hntool
.RB [ " \-ln " ]
.SH DESCRIPTION
.I hntool
Hardening tool for Unix. It scans the system and available software, 
to detect security.

.SH OPTIONS
.TP
.B \-l --list
returns list of available rules
.TP
.B \-n --nocolors
does not use colors on output
.SH "UNDERSTANDING THE OUTPUT"
.TP
There are 5 types of results:
.TP
.B OK : 
  Means that the item checked is fine and that you do not need to worry
.TP
.B INFO: 
  Means that you should know the item status, but probably it is fine. A port
  opened, for example.
.TP
.B LOW:
  Means that a security problem was found, but it does not provides a high risk
  for your system.
.TP
.B MEDIUM:
  Things are getting worse and you should start to worry about these itens.
.TP
.B HIGH:
  You have an important security hole/problem on your system and you
  should fix it NOW or run and save your life.
.SH "COPYRIGHT NOTICE"
Copyright \(co 1998, 1999, 2001, 2002 Free Software Foundation, Inc.
.br
Copyright \(co 1992, 1993 Jean-loup Gailly
.PP
Permission is granted to make and distribute verbatim copies of
this manual provided the copyright notice and this permission notice
are preserved on all copies.
.ig
Permission is granted to process this file through troff and print the
results, provided the printed document carries copying permission
notice identical to this one except for the removal of this paragraph
(this paragraph not being relevant to the printed manual).
..
.PP
Permission is granted to copy and distribute modified versions of this
manual under the conditions for verbatim copying, provided that the entire
resulting derived work is distributed under the terms of a permission
notice identical to this one.
.PP
Permission is granted to copy and distribute translations of this manual
into another language, under the above conditions for modified versions,
except that this permission notice may be stated in a translation approved
by the Foundation.


hntool.py
#! /usr/bin/env python
#
# hntool - A hardening tool for Linux/BSD
# Copyright (C) 2009 Hugo Doria <mail@hugodoria.org>
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#

import hntool.util
import getopt
import hntool
import string
import sys


# Return all possible modules (rules)
def get_modules():
	'''Method to return all modules available'''
	return hntool.__all__

# vars
use_colors = True # using colors by default
hntool_version = 0.1

# yes, only unix for now
if not hntool.util.is_unix():
	print 'Error: You must have a Unix(-like) box. (No candy for you)'
	sys.exit(2)

# checking if we are root. we need to be. oh yeah, baby.
if not hntool.util.is_root():
	print 'Error: You must be root to run hntool'
	print
	print hntool.util.usage()
	sys.exit(2)

# get our options and process them
try:
	optlist, args = getopt.getopt(sys.argv[1:], "lhn", ["list","help","nocolors"])
except getopt.GetoptError:
	hntool.util.usage()

for i, k in optlist:
	# Show all available modules (rules) and its description
	if i in ('-l','--list'):
		print "-"*20 + " hntool rule list " + "-"*20

		for module in get_modules():
			print string.ljust(module, 20) + ": " + __import__('hntool.' + \
					module, globals(), locals(), [hntool]).rule().long_name()

		sys.exit(2)

	# do not use colors
	if i in ('-n', '--nocolors'):
		use_colors = False

	# show help
	if i in ('-h', '--help'):
		hntool.util.usage()


# Method to show the check results
def msg_status(msg, status):
	'''
	Method to show the check results
	'''
	if use_colors:
		if status == 'ok':
			return string.ljust(msg,70) + hntool.util.color_ok()
		elif status == 'low':
			return string.ljust(msg,70) + hntool.util.color_low()
		elif status == 'medium':
			return string.ljust(msg,70) + hntool.util.color_medium()
		elif status == 'high':
			return string.ljust(msg,70) + hntool.util.color_high()
		elif status == 'info':
			return string.ljust(msg,70) + hntool.util.color_info()

	else:
		if status == 'ok':
			return string.ljust(msg,70) + '[   OK   ]'
		elif status == 'low':
			return string.ljust(msg,70) + '[  LOW   ]'
		elif status == 'medium':
			return string.ljust(msg,70) + '[ MEDIUM ]'
		elif status == 'high':
			return string.ljust(msg,70) + '[  HIGH  ]'
		elif status == 'info':
			return string.ljust(msg,70) + '[  INFO  ]'	


print '[ Starting hntool checks ]'

# Run all the modules and its checks. The results of each module goes to "check_results"
for module in get_modules():
	check_results = [[],[],[],[],[]] # ok, low, medium, high and info

	# Getting all results from the check and the module/check description
	check_results = __import__('hntool.' + module ,globals(), locals(), [hntool]).rule().analyze()
	check_description = __import__('hntool.' + module ,globals(), locals(), [hntool]).rule().long_name()

	# Prints the module description
	print '\n' + check_description

	# Print all the results, from the 5 types of messages (ok, low, medium, high and info).
	# First message is the "ok" one (check_results[0]). The second one is
	# "low" (check_results[1]). The third (check_results[2]) is for "warnings"
	# and the fourth one is "high" (check_results[3]), The last one is for
	# info messages.
	if check_results[0] != []:
		for j in check_results[0]:
			print '   ' + msg_status(j, 'ok')
	if check_results[1] != []:
		for j in check_results[1]:
			print '   ' + msg_status(j, 'low')
	if check_results[2] != []:
		for j in check_results[2]:
			print '   ' + msg_status(j, 'medium')
	if check_results[3] != []:
		for j in check_results[3]:
			print '   ' + msg_status(j, 'high')
	if check_results[4] != []:
		for j in check_results[4]:
			print '   ' + msg_status(j, 'info')
